ouroboros-network/src/Ouroboros/Network/Server/ConnectionTable.hs

```diff
@@ -0,0 +1,214 @@
+
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Ouroboros.Network.Server.ConnectionTable (
+      ConnectionTable
+    , ConnectionTableRef (..)
+    , ValencyCounter
+
+    , newConnectionTable
+    , refConnection
+    , addConnection
+    , removeConnection
+    , newValencyCounter
+    , addValencyCounter
+    , remValencyCounter
+    , waitValencyCounter
+    , readValencyCounter
+    ) where
+
+import           Control.Monad (when)
+import           Control.Monad.Class.MonadSTM
+--import           Control.Tracer XXX Not Yet
+import qualified Data.Map.Strict as M
+import           Data.Set (Set)
+import qualified Data.Set as S
+import qualified Network.Socket as Socket
+import           Text.Printf
+
+-- A ConnectionTable represent a set of connections that is shared between
+-- servers and subscription workers. Its main purpose is to avoid the creation of duplicate
+-- connections (especially connections with identical source address, destination address, source
+-- port and destination port which would be rejected by the kernel anyway.).
+-- It is only used for bookkeeping, the sockets represented by the connections are not accessable
+-- through this structure.
+data ConnectionTable m = ConnectionTable {
+    ctTable     :: TVar m (M.Map Socket.SockAddr (ConnectionTableEntry m))
+  , ctLastRefId :: TVar m Int
+  }
+
+-- | ValencyCounter represents how many active connections we have towards a given peer.
+-- It starts out with a positive value representing a desired number of connections for a specific
+-- subscription worker. It can become negative, for example if a peer opens multiple connections
+-- to us.
+-- The vcId is unique per ConnectionTable and ensures that we won't count the same connection twice.
+data ValencyCounter m = ValencyCounter {
+    vcId  :: Int
+  , vcRef :: TVar m Int
+  }
+
+-- | Create a new ValencyCounter
+newValencyCounter
+  :: MonadSTM m
+  => ConnectionTable m
+  -> Int
+  -- ^ Desired valency, that is number of connections a subscription worker will attempt to
+  -- maintain.
+  -> STM m (ValencyCounter m)
+newValencyCounter tbl valency =  do
+  lr <- readTVar $ ctLastRefId tbl
+  let !lr' = lr + 1
```

Comment:
Alternatively, we could put this bang in `ValencyCounter`.  It will also force it here.

Reply:
Will leave it as is for now.